refactor account related methods & add support for MintTokensTx

jordipainan opened the pull request.

jordipainan committed:
ethless: add support for mintTokensTx

jordipainan committed:
ethless: update setAccountInfo and account methods:

- SetAccountInfoTx now supports to create an account
consuming a previously generated faucet payload
- Splits the account creation and costs and nonce updates
- The cost of a tx is sended to the burn address
- Adds checks for delegates, nonces and transaction cost
- Adds Client methods for get/set account
- Adds Client method for generating a faucet payload
- Adds integration tests for setAccountInfoURI txs

Adds more unit tests for account creation

mvdan commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -1059,3 +1077,136 @@ func testSetTxCost(mainClient *client.Client, treasurerSigner *ethereum.SignKeys
+			log.Warnf("error getting current height: %v", err)
+			continue
+		} else {
+			if h2 > h {
```
nit: join else and if

mvdan commented on cmd/vochaintest/vochaintest.go:
```diff
@@ -1059,3 +1077,136 @@ func testSetTxCost(mainClient *client.Client, treasurerSigner *ethereum.SignKeys
+			log.Warnf("error getting current height: %v", err)
+			continue
+		} else {
+			if h2 > h {
```
nit: same

mvdan commented on vochain/account.go:
```diff
@@ -192,94 +171,406 @@ func (v *State) VerifyAccountBalance(message, signature []byte, amount uint64) (
+) error {
 	// check valid address
-	if address.String() == types.EthereumZeroAddress {
+	if bytes.Equal(accountAddress.Bytes(), types.EthereumZeroAddressBytes[:]) {
```
FYI, you can replace all instances of `bytes.Equal(foo.Bytes(), types.EthereumZeroAddressBytes[:])` with just `foo == types.EthereumZeroAddressBytes` - you can compare array values directly.

mvdan commented on types/consts.go:
```diff
@@ -7,8 +7,9 @@ func Bool(b bool) *bool { return &b }
-	True  = Bool(true)
+	False                    = Bool(false)
+	True                     = Bool(true)
+	EthereumZeroAddressBytes = [20]byte{}
```
nit: I'd do `EthereumZeroAddress ethcommon.Address`.

This drops the `Bytes`, because it's a byte array, not a byte slice, and "bytes" tends to imply a slice.

you also give it the proper type of Address, which helps avoid misuse, and also allows you to not hard-code the length of 20 here.

mvdan commented on vochain/account.go:
```diff
@@ -192,94 +171,406 @@ func (v *State) VerifyAccountBalance(message, signature []byte, amount uint64) (
+	}
+	// check nonce
+	if tx.Nonce != treasurer.Nonce {
+		return common.Address{}, 0, fmt.Errorf("invalid nonce %d, expected: %d", tx.Nonce, treasurer.Nonce+1)
```
you check against Nonce but report Nonce+1?

mvdan commented on vochain/account.go:
```diff
@@ -192,94 +171,406 @@ func (v *State) VerifyAccountBalance(message, signature []byte, amount uint64) (
+					returnValues.FaucetPayloadSigner = issuerAddress
+					returnValues.FaucetPayload = tx.GetFaucetPackage().Payload
+				} else {
+					return nil, fmt.Errorf("payload to and tx sender missmatch")
```
tip: if you reverse this if and return early, then you can unindent the bulk of the code. same for some other "ifs" that have this pattern.

mvdan commented on vochain/account.go:
```diff
@@ -136,26 +123,18 @@ func (v *State) MintBalance(address common.Address, amount uint64) error {
-		}
+	acc, err := v.GetAccount(address, false)
+	if err != nil {
+		return fmt.Errorf("mintBalance (%w)", err)
```
nit: the format is `foo: %w` for %w to work.

mvdan commented on vochain/account.go:
```diff
@@ -87,46 +87,33 @@ func (a *Account) DelDelegate(addr common.Address) {
+func (v *State) TransferBalance(from, to common.Address, amount uint64, nonce uint32) error {
+	var accFrom, accTo *Account
+	var err error
+	if accFrom, err = v.GetAccount(from, false); err != nil {
```
You don't need these `var`s:

```suggestion
	accFrom, err := v.GetAccount(from, false)
	if err != nil {
```

mvdan replied:
joining the error check with the `if` can help reduce lines in some situations, but in this case it just doesn't help

mvdan reviewed (commented):
quick review, happy to be more thorough when it's no longer WIP

jordipainan committed:
address suggestions:

Also reduce block time to 3 seconds on tests
and wait more when fetching current block,
thus ensuring that the block was propagated
and required transactions are commited
into the persistent state

randomshinichi reviewed (approved):
LGTM
